src/layers/attention.jl

```diff
@@ -1,58 +1,48 @@
-"""
-struct Attention{T}
-  qkv::T
-end
-
-Attention(dims::Pair{Int, Int}) = Attention(Dense(dims.first, dims.second * 3; bias = false))
-
-@functor Attention
-
-function (attn::Attention)(x::AbstractArray{T, 3}) where T
-  q, k, v = chunk(attn.qkv(x), 3, dims = 1)
-  scale = convert(T, sqrt(size(q, 1)))
-  score = softmax(batched_mul(batched_transpose(q), k) / scale)
-  attention = batched_mul(v, score)
+Multi-head self-attention layer.
 
-  return attention
-end
-
-struct MHAttention{S, T}
-  heads::S
-  projection::T
+# Arguments:
+- `nheads`: Number of heads
+- `qkv_layer`: layer to be used for getting the query, key and value
+- `attn_drop`: dropout rate after the self-attention layer
+- `projection`: projection layer to be used after self-attention
+"""
+struct MHAttention{P, Q, R}
+  nheads::Integer
+  qkv_layer::P
+  attn_drop::Q
+  projection::R
 end
 
 """
-    MHAttention(in, hidden, nheads; dropout = 0.0)
+    MHAttention(planes, nheads = 8; qkv_bias = false, attn_drop = 0., proj_drop = 0.)
 
-Multi-head self-attention layer used in many vision transformer-like models.
+Multi-head self-attention layer.
 
-# Arguments
-- `in`: Number of dimensions in the input.
-- `hidden`: Number of dimensions in the intermediate layer.
-- `nheads`: Number of attention heads.
-- `dropout`: Dropout rate for the projection layer.
+# Arguments:
+- `planes`: number of input channels
+- `nheads`: number of heads
+- `qkv_bias`: whether to use bias in the layer to get the query, key and value
+- `attn_drop`: dropout rate after the self-attention layer
+- `proj_drop`: dropout rate after the projection layer
 """
-function MHAttention(in, hidden, nheads; dropout = 0.)
-  if (nheads == 1 && hidden == in)
-    return Attention(in => in)
-  end
-  inheads, innerheads = chunk(1:in, nheads), chunk(1:hidden, nheads)
-  heads = Parallel(vcat, [Attention(length(i) => length(o)) for (i, o) in zip(inheads, innerheads)]...)
-  projection = Chain(Dense(hidden, in), Dropout(dropout))
-
-  MHAttention(heads, projection)
+function MHAttention(planes, nheads = 8; qkv_bias = false, attn_drop = 0., proj_drop = 0.)
+  @assert planes % nheads == 0 "planes should be divisible by nheads"
+  qkv_layer = Dense(planes, planes * 3; bias = qkv_bias)
+  attn_drop = Dropout(attn_drop)
+  proj = Chain(Dense(planes, planes), Dropout(proj_drop))
+
+  MHAttention(nheads, qkv_layer, attn_drop, proj)
 end
 
 @functor MHAttention
 
-function (mha::MHAttention)(x)
-  nheads = length(mha.heads.layers)
-  xhead = chunk(x, nheads, dims = 1)
-  return mha.projection(mha.heads(xhead...))
+function (m::MHAttention)(x::AbstractArray{T, 3}) where T
+  B, C, N = size(x)
+  q, k, v = chunk(reshape(m.qkv_layer(x), B ÷ m.nheads, m.nheads, C, 3 * N), 3; dims = 4)
+  scale = convert(T, sqrt(size(q, 1) / m.nheads))
+  attn = m.attn_drop(softmax(NeuralAttentionlib.matmul(q, permutedims(k, (2, 1, 3, 4))) * scale))
+  x = m.projection(reshape(NeuralAttentionlib.matmul(attn, v), (B, C, N)))
```

Comment:
Out of curiosity, why not use `NeuralAttentionlib.multihead_qkv_attention`?

Reply:
Primarily for demonstration purposes....here of course the library function works perfectly well because it's vanilla attention but for other models there are more things happening and so I wrote it this way to kinda document the internals of self attention for a ViT-like model as more of these models are added. Could always switch - is there a significant performance boost obtained?

Reply:
>  is there a significant performance boost obtained?

Probably no, since they work in the same way. But I wonder where does the performance gap (~ 100ms) you mentioned coming from. The functions `NeuralAttentionlib` (currently) provided is to make attention functions composable, and `matmul` should be equivalent to `NNlib.batch_mul` + `reshape`. I'm not sure if the performance gain really comes from `NeuralAttentionlib`

Reply:
Maybe it's the multiple calls to `Attention` inside `Parallel`? This one simply does MHA directly while the original implementation had `Parallel` with multiple `Attentions` zipped together. The main reason I chose this version was because it also leaves future options open for working with inputs with more than three dimensions if need be (as I understand it `CollapsedDimArray` provides this functionality if necessary)

Reply:
> it also leaves future options open for working with inputs with more than three dimensions if need be (as I understand it CollapsedDimArray provides this functionality if necessary

Yes, and actually `NeuralAttentionlib.multihead_qkv_attention` already allows inputs with more dimensions as long as the dimension order is correct.

> while the original implementation had Parallel with multiple Attentions zipped together. 

So this seems to be the main reason for the performance gap. `Parallel` doesn't run in parallel, it sequentially execute each path and aggregate the result, so doing MHA directly would definitely be faster.

Reply:
As I recall in the original PR had issues with `batched_mul` across channels, and it also performed the multiplications serially. Does `NeuralAttentionlib` parallelize them?

Reply:
https://github.com/chengchingwen/NeuralAttentionlib.jl/blob/master/src/matmul/matmul.jl certainly seems to work similar to the NNlib version (but with options for more dimensions), so I think there is parallelism? @chengchingwen could elaborate better though

Reply:
I meant #105 for reference. In that PR, 4D `batched_mul` was originally implemented as a sequential loop over the 3rd dimension (which ended up being faster via `Parallel`). I thought there was already some reshaping going on, but maybe we missed doing it correctly.

Either way, looks like at least the GPU path is parallelized for sure. It's good to consolidate these primitives in NeuralAttentionlib.jl. In #105, part of the problem with 4D `batched_mul` was making assumptions about what the different dimensions mean. In NeuralAttentionlib.jl, it will be much safer to do that.